### ISSUE TITLE
Investigate read and read_batch performance degradation at large symbol counts

### ISSUE THREAD

**Issue by G-D-Petrov** [man-group/ArcticDB#869]
### Describe the bug

We have noticed that the performance of read_batch method of the VersionStore degrades when called with large batches of symbols. More specifically the performance scales super linearly with the number of symbols. For example:

  | 1000 | 5000 | 10000
-- | -- | -- | --
Duration | 4.62 | 47.73 | 184.35
Avg Time per Symbol | 0.0046 | 0.00954 | 0.018435

The problem is that the average time for processing a symbol is increasing significantly, which is not expected.
This is not experienced with regular, sequential reads. This leads to sequential reads being faster for large enough sets of symbols.
Example for the performance of sequential reads over the same symbols/data can be seen below:

  | 1000 | 5000 | 10000
-- | -- | -- | --
Duration | 18.45 | 95.56 | 203.75
Avg Time per Symbol | 0.01845 | 0.019 | 0.02

At around 10000 symbols, the performance of sequential reads is almost the same as batch reads, for the same underlying symbols/data.

### Steps/Code to Reproduce

``` python
from arcticdb import Arctic
from arcticdb.version_store.library import ReadRequest
import pandas as pd
import time

ac = Arctic()
lib = ac['some_library']

syms = lib.list_symbols()
syms.sort()
COLS = [
    "OPEN", 
    "HIGH", 
    "LOW", 
    "CLOSE"
]

dr = pd.date_range(start='2020-12-16', end = '2020-12-18', freq='T')

for size in [1000, 5000, 10000]:
    read_reqs = [ReadRequest(f"{sym}_sym", date_range=dr, columns=COLS) for sym in range(size)]
    _ = lib .read_batch(read_reqs)
```

### Expected Results

The total time for the read_batch should scale linearly(or close to it) and not super linearly as now. 

### OS, Python Version and ArcticDB Version

Any version

### Backend storage used

Any storage

**Comment by G-D-Petrov** [comment]
TLDR: It looks like the major problem is with this line:
https://github.com/man-group/ArcticDB/blob/master/cpp/arcticdb/pipeline/python_output_frame.cpp#L53

and the fact that is is calling malloc_trim on Linux, this is causing a major bottleneck at the end of batch_read (see comparison below)
This is _also causing some delays for individual reads_, but not as pronounced.

We can distinguish 3 major parts to read_batch:
- the actual reading from storage - this is mostly done asynchronously 
- post_processing of the data frame - this is done sequentially for each read result/symbol
- deallocation of the PythonOutput frame - also done sequentially at the end when the ReadResult vector is dropped

Below you can see a table that illustrated how much time(in sec) is spent on each step for given number of symbols.
The times are not entirely accurate as they were taken while the program was profiled with py-spy, but they are still representative.

  | 1000 | 5000 | 10000 | 15000
-- | -- | -- | -- | --
read | 4.2 | 24.5 | 76 | 69
post_process | 5.7 | 42 | 84 | 100
dealoc | 3.7 | 160 | 900 | 1320
total | 13.6 | 226.5 | 1060 | 1489

As you can see most of the time is spent on deallocation.
This is because for each ReadResult, there is a PythonOutputFrame for which we are calling malloc_trim and this is taking a lot of time.
Below I have compiled a similar tables but with a version that has the line that calls malloc_trim commented out and with regular executions.
(All tables below are without a profiler, so the times are faster than the table above)

**Batch - with malloc_trim**
  | 1000 | 5000 | 10000 | 15000
-- | -- | -- | -- | --
Duration | 12 | 103 | 339 | 650
Avg Time per Symbol | 0.012 | 0.0206 | 0.0339 | 0.043333

**Batch - without malloc_trim**
  | 1000 | 5000 | 10000 | 15000 | 20000 | 30000 | 40000 | 50000 | 60000
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
Duration | 2.37 | 11.67 | 26.85 | 46.5 | 65.72 | 116.55 | 193.83 | 286.84 | 362.02
Avg Time per Symbol | 0.00237 | 0.002334 | 0.002685 | 0.0031 | 0.003286 | 0.003885 | 0.004846 | 0.005737 | 0.006034

**As you can see, the version without malloc_trim is about 5-15 times faster.**
The post-processing is still somewhat of a bottleneck(it is not running asynchronously) , but not nearly to the same extent as the malloc_trim.
I also ran the same experiments for **sequential** reads -  there the difference is much smaller, but there are still cases where the version **without malloc_trim is up-to 10% faster**.

**Sequential - with malloc_trim**
  | 1000 | 5000 | 10000 | 15000
-- | -- | -- | -- | --
Duration | 21.86 | 103.5 | 182 | 266
Avg Time per Symbol | 0.02186 | 0.0207 | 0.0182 | 0.017733


**Sequential - without malloc_trim**
  | 1000 | 5000 | 10000 | 15000
-- | -- | -- | -- | --
Duration | 21.37 | 91.65 | 173 | 261.44
Avg Time per Symbol | 0.02137 | 0.01833 | 0.0173 | 0.017429

Removing malloc_trim doesn't seem to have affected the memory usage.
Even on my stress test that went from 1000 up to 60000 symbols, in a continuous loop, **there were no noticeable memory issues**.

**I am in favour of simple removing [this line](https://github.com/man-group/ArcticDB/blob/master/cpp/arcticdb/pipeline/python_output_frame.cpp#L53) from the PythonOutputFrame destructor, it is already working only on Linux (Windows doesn't have an equvalent of malloc_trim) and we haven't had problems on Windows., as far as I am aware.**
We can expose a new method for calling malloc_trim on Linux and that method should be called only once.
